File: [J]backtracking/[J]backtracking/17-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        res = []
        num_chars = {'2': 'abc', '3': 'def', '4':'ghi', '5':'jkl', '6':'mno', '7':'pqrs', '8':'tuv', '9':'wxyz'}

        def dfs(path, idx):
            if len(path) == len(digits):
                res.append(''.join(path))
                return
            for c in num_chars[digits[idx]]:
                path.append(c)
                dfs(path, idx + 1)
                path.pop()

        dfs([], 0)
        return res
                
# time O(4**n), n is the length of digits
# space O(n), due to recursion stack, output is O(4**n)
# using dfs and backtracking and backtracking with constraints

File: [J]backtracking/[J]backtracking/17-letter-combinations-of-a-phone-number.java
class Solution {

    HashMap<Character, ArrayList<Character>> map;
    List<String> res;

    public List<String> letterCombinations(String digits) {
        map = new HashMap<>();
        int cur = 0;
        for (int i = 2; i < 10; i++) {
            char c = (char) (i + '0');
            int total = 3;
            if (i == 7 || i == 9) {
                total += 1;
            }
            for (int j = 0; j < total; j++) {
                map.computeIfAbsent(c, k -> new ArrayList<>()).add((char) (cur + 'a'));
                cur += 1;
            }
        }
        res = new ArrayList<>();
        if (digits.length() > 0) {
            dfs(digits, new StringBuffer(), 0);
        }
        return res;
    }

    public void dfs(String digits, StringBuffer sb, int idx) {
        if (idx == digits.length()) {
            res.add(sb.toString());
            return;
        }
        for (int i = 0; i < map.get(digits.charAt(idx)).size(); i++) {
            Character c = map.get(digits.charAt(idx)).get(i);
            sb.append(c);
            dfs(digits, sb, idx + 1);
            sb.deleteCharAt(sb.length() - 1);
        }
        return;
    }
}

// time O(4**n), n is the length of digits
// space O(n), due to recursion stack, output is O(4**n)
// using dfs and backtracking and backtracking with constraints